runtime/oti/VMHelpers.hpp

```diff
@@ -756,7 +756,7 @@ class VM_VMHelpers
 #if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
 		if (J9_IS_J9CLASS_FLATTENED(arrayClass)) {
 			instance = objectAllocate->inlineAllocateIndexableValueTypeObject(currentThread, arrayClass, size, initializeSlots, memoryBarrier, sizeCheck);
-		} else
+		} else if (J9_ARE_NO_BITS_SET(arrayClass->classFlags, J9ClassContainsUnflattenedFlattenables))
```

Comment:
I'd like to see TraceAssert added to `inlineAllocateIndexableObject` and similar helpers to validate that we're not passing any classes that have the `J9ClassContainsUnflattenedFlattenables` bit set.

The assert should be guarded by `J9VM_OPT_VALHALLA_VALUE_TYPES`.  FYI @tajila

Reply:
Looking at the interpreter code, I think we have different semantics there and may trip up on a case like:
```java
inline class I {
   int a;
   int b;
}

class Bad {
  I i;
  int c;
}

new Bad[10];
```
See 
https://github.com/eclipse/openj9/blob/c6b594b6c03f957bb6a07f5d234a6d996037a17f/runtime/vm/BytecodeInterpreter.hpp#L1079-L1087
which calls:
https://github.com/eclipse/openj9/blob/c6b594b6c03f957bb6a07f5d234a6d996037a17f/runtime/vm/BytecodeInterpreter.hpp#L1051-L1061

which means either the inline allocate code already handles this or the interpreter needs to be fixed too

Reply:
Based on @tajila's response, the code I missed is in the `anewarray` implementation:
https://github.com/eclipse/openj9/blob/c6b594b6c03f957bb6a07f5d234a6d996037a17f/runtime/vm/BytecodeInterpreter.hpp#L7569-L7574